In [1]:
import pandas as pd
import numpy as np

granify_data = pd.read_csv(r"C:\Users\level\Downloads\granifyproject-main\granifyproject-main\BA Candidate Assignment 1 Files [EXTERNAL]\granify_orders.csv")
crisskross_data = pd.read_csv(r"C:\Users\level\Downloads\granifyproject-main\granifyproject-main\BA Candidate Assignment 1 Files [EXTERNAL]\crisskross_orders.csv")

crisskross_data = crisskross_data.rename(columns={"Revenue" : "Revenue_ck", "Segment" : "Segment_ck"})

pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', -1)

print(granify_data.head())
print(crisskross_data.head())

   site_id   order_id                 date  test_group  campaign_id  subtotal  \
0  1402     367521923  2020-02-07 05:00:25  0.0        NaN           155.08     
1  1402     367521967  2020-02-07 05:00:25  0.0        NaN           155.08     
2  1402     367522017  2020-02-07 05:00:25  0.0        NaN           70.99      
3  1402     367522077  2020-02-07 05:00:25  0.0        NaN           70.99      
4  1402     367522615  2020-02-07 05:00:25  0.0        NaN           38.99      

  traffic_type             inserted   os os_version browser  browser_version  \
0  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              
1  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              
2  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              
3  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              
4  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              

  device_category               

In [2]:
combined_data = granify_data.join(crisskross_data.set_index("Transaction ID"), on="order_id")

print(combined_data.head())

   site_id   order_id                 date  test_group  campaign_id  subtotal  \
0  1402     367521923  2020-02-07 05:00:25  0.0        NaN           155.08     
1  1402     367521967  2020-02-07 05:00:25  0.0        NaN           155.08     
2  1402     367522017  2020-02-07 05:00:25  0.0        NaN           70.99      
3  1402     367522077  2020-02-07 05:00:25  0.0        NaN           70.99      
4  1402     367522615  2020-02-07 05:00:25  0.0        NaN           38.99      

  traffic_type             inserted   os os_version browser  browser_version  \
0  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              
1  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              
2  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              
3  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              
4  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              

  device_category               

In [3]:
#Flags rows with discrepancies in revenue
combined_data["revenue_flag"] = np.where(combined_data["subtotal"] != combined_data["Revenue_ck"], 1, 0)

print(combined_data.head())

   site_id   order_id                 date  test_group  campaign_id  subtotal  \
0  1402     367521923  2020-02-07 05:00:25  0.0        NaN           155.08     
1  1402     367521967  2020-02-07 05:00:25  0.0        NaN           155.08     
2  1402     367522017  2020-02-07 05:00:25  0.0        NaN           70.99      
3  1402     367522077  2020-02-07 05:00:25  0.0        NaN           70.99      
4  1402     367522615  2020-02-07 05:00:25  0.0        NaN           38.99      

  traffic_type             inserted   os os_version browser  browser_version  \
0  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              
1  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              
2  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              
3  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              
4  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              

  device_category               

In [4]:
#This is to check to verify that 1 denotes a baseline group member. If I sum the column and the sum is roughly 10% of the number
#of entries, then 1 does in fact denote the baseline group.
print(np.sum(combined_data["test_group"]))

combined_data.info()

12499.0
<class 'pandas.core.frame.DataFrame'>
Int64Index: 128323 entries, 0 to 127199
Data columns (total 17 columns):
site_id            128323 non-null int64
order_id           128323 non-null int64
date               128323 non-null object
test_group         128198 non-null float64
campaign_id        55771 non-null float64
subtotal           128323 non-null float64
traffic_type       128323 non-null object
inserted           128323 non-null object
os                 128323 non-null object
os_version         108914 non-null object
browser            128323 non-null object
browser_version    123340 non-null float64
device_category    128323 non-null object
ssid               127848 non-null object
Segment_ck         123206 non-null object
Revenue_ck         123206 non-null float64
revenue_flag       128323 non-null int32
dtypes: float64(5), int32(1), int64(2), object(9)
memory usage: 17.1+ MB


And we can see that 12,499 rows have a "test_group" value of 1, which is roughly 10% of the 127,199 total rows.

In [5]:
#Numeric version of Segement_CK so that it can easily be compared with test_group
combined_data["segment_num"] = np.where(combined_data["Segment_ck"] == "granify", 0, 1)

print(combined_data.head())

#Flags rows with discrepancies in orders
combined_data["order_flag"] = np.where(combined_data["test_group"] != combined_data["segment_num"], 1, 0)

   site_id   order_id                 date  test_group  campaign_id  subtotal  \
0  1402     367521923  2020-02-07 05:00:25  0.0        NaN           155.08     
1  1402     367521967  2020-02-07 05:00:25  0.0        NaN           155.08     
2  1402     367522017  2020-02-07 05:00:25  0.0        NaN           70.99      
3  1402     367522077  2020-02-07 05:00:25  0.0        NaN           70.99      
4  1402     367522615  2020-02-07 05:00:25  0.0        NaN           38.99      

  traffic_type             inserted   os os_version browser  browser_version  \
0  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              
1  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              
2  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              
3  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              
4  active       2020-02-08 23:49:40  Mac  OS X       Safari  13.0              

  device_category               

In [43]:
order_discrepancies = np.sum(combined_data["order_flag"])
revenue_discrepancies = np.sum(combined_data["revenue_flag"])
total_orders = len(combined_data.index)

combined_data["discrepancy_flag"] = np.where((combined_data["revenue_flag"] == 1) | (combined_data["order_flag"] == 1), 1, 0)

any_discrepancy = np.sum(combined_data["discrepancy_flag"])

print("The number of rows with revenue discrepancies is", revenue_discrepancies, "or", "{:%}".format(revenue_discrepancies/total_orders))
print("The number of rows with order discrepancies is", order_discrepancies, "{:%}".format(order_discrepancies/total_orders))
print("The number of rows with at least one discrepancy is", any_discrepancy, "{:%}".format(any_discrepancy/total_orders))

The number of rows with revenue discrepancies is 5131 or 3.998504%
The number of rows with order discrepancies is 5135 4.001621%
The number of rows with at least one discrepancy is 5612 4.373339%


Ok, so there are discrepancies about 4% of the time on both revenue and orders. While not terrible, that's not great either. As per the instructions, Granify prefers the delta to be under 3%.

First I'll investigate the discrepancies between order type. Below I'm playing around with the data to see if there are any patterns in the flagged orders versus all orders.

In [7]:
#Creates a dataframe that contains only rows with order discrepancies
flagged_orders = combined_data[combined_data["order_flag"] == 1]
non_flagged_orders = combined_data[combined_data["order_flag"] == 0]

print(flagged_orders.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5135 entries, 22 to 127199
Data columns (total 19 columns):
site_id            5135 non-null int64
order_id           5135 non-null int64
date               5135 non-null object
test_group         5010 non-null float64
campaign_id        2530 non-null float64
subtotal           5135 non-null float64
traffic_type       5135 non-null object
inserted           5135 non-null object
os                 5135 non-null object
os_version         4434 non-null object
browser            5135 non-null object
browser_version    4800 non-null float64
device_category    5135 non-null object
ssid               4997 non-null object
Segment_ck         481 non-null object
Revenue_ck         481 non-null float64
revenue_flag       5135 non-null int32
segment_num        5135 non-null int32
order_flag         5135 non-null int32
dtypes: float64(5), int32(3), int64(2), object(9)
memory usage: 742.2+ KB
None


In [8]:
print("{:%}".format(np.sum(flagged_orders["revenue_flag"])/len(flagged_orders.index)), "of order discrepancies also had a revenue discrpency.")


90.632911% of order discrepancies also had a revenue discrpency.


In [9]:
print(flagged_orders["os"].value_counts())
print("\n", flagged_orders["browser"].value_counts())
print("\n", flagged_orders["device_category"].value_counts())
print("\n", flagged_orders["test_group"].value_counts())
print("\n", flagged_orders["campaign_id"].value_counts())

iOS        1635
Windows    1526
Android    1321
Mac        589 
Other      50  
Linux      14  
Name: os, dtype: int64

 Chrome                 956
Mobile Chrome          912
Mobile Safari (App)    844
Mobile Safari          695
Firefox                584
Safari                 436
other                  335
Edge                   265
Samsung Browser        101
Facebook               5  
IE                     1  
Opera                  1  
Name: browser, dtype: int64

 mobile     2245
desktop    2133
tablet     692 
unknown    65  
Name: device_category, dtype: int64

 0.0    4814
1.0    196 
Name: test_group, dtype: int64

 6230.0    642
6733.0    632
6727.0    584
6381.0    120
6728.0    93 
6732.0    82 
6552.0    80 
6190.0    78 
6600.0    53 
6765.0    51 
6764.0    22 
6924.0    20 
6606.0    15 
6752.0    13 
6763.0    12 
6888.0    10 
6889.0    10 
6762.0    9  
6631.0    3  
6749.0    1  
Name: campaign_id, dtype: int64


In [10]:
func1 = lambda x: 100*x.count()/flagged_orders.shape[0]
func2 = lambda x: 100*x.count()/non_flagged_orders.shape[0]

flagged_os_check = pd.pivot_table(flagged_orders, index =['os', "os_version"], values="order_flag", aggfunc=func1)
combined_os_check = pd.pivot_table(non_flagged_orders, index =['os', "os_version"], values="order_flag", aggfunc=func2)

os_check = flagged_os_check.join(combined_os_check, lsuffix="_flagged", rsuffix="_nf")
os_check["diff"] = os_check["order_flag_flagged"] - os_check["order_flag_nf"]
display(os_check.sort_values("diff"))




order_flag_flagged  order_flag_nf      diff
os      os_version                                             
iOS     13          21.694255           26.972595     -5.278340
Mac     OS X        11.470302           13.357632     -1.887330
iOS     10          0.662123            0.834497      -0.172374
        9           0.350536            0.496802      -0.146266
        11          0.973710            1.039062      -0.065352
Android 4.0         0.077897            0.123389      -0.045492
        4.2         0.019474            0.023541      -0.004067
iOS     7           0.019474            0.011365       0.008109
        8           0.038948            0.025977       0.012972
Android 4.1         0.038948            0.021918       0.017031
        4.3         0.019474            0.000812       0.018662
Windows XP          0.194742            0.107965       0.086777
Android 4.4         0.408958            0.292236       0.116722
Windows 8           0.253165            0.087671       0.165494
Android 6           1.226874            1.056109       0.170765
Windows Vista       0.350536            0.173718       0.176817
Android 7           3.563778            3.338799       0.224979
        8           4.673807            4.327532       0.346275
Windows 8.1         1.421616            1.022015       0.399601
Android 5           3.291139            2.794103       0.497036
iOS     12          8.081792            6.444621       1.637170
Windows 7           4.634859            2.442608       2.192251
        10          22.745862           19.817677      2.928185
        Phone       0.116845           NaN            NaN      
iOS     3           0.019474           NaN            NaN

In [11]:
flagged_browser_check = pd.pivot_table(flagged_orders, index =['browser', "browser_version"], values="order_flag", aggfunc=func1)
combined_browser_check = pd.pivot_table(non_flagged_orders, index =['browser', "browser_version"], values="order_flag", aggfunc=func2)

browser_check = flagged_browser_check.join(combined_browser_check, lsuffix="_flagged", rsuffix="_nf")
browser_check["diff"] = browser_check["order_flag_flagged"] - browser_check["order_flag_nf"]
display(browser_check.sort_values("diff"))

browser_check.to_csv(r"C:\Users\level\Downloads\granifyproject-main\granifyproject-main\BA Candidate Assignment 1 Files [EXTERNAL]\order_discrepancy_browser.csv")

order_flag_flagged  order_flag_nf  \
browser             browser_version                                      
Mobile Safari       13.0             7.283350            10.628470       
Safari              13.0             7.575463            10.207975       
Edge                18.0             4.926972            7.372471        
Mobile Chrome       79.0             9.464460            10.981589       
Mobile Safari (App) 13.0             13.106134           14.352047       
Mobile Chrome       80.0             5.335930            6.211644        
                    54.0             0.759494            1.252557        
Chrome              80.0             5.043817            5.354418        
Mobile Chrome       78.0             0.428432            0.721661        
Safari              12.0             0.214216            0.497613        
Mobile Safari       11.0             0.486855            0.765497        
Facebook            256.0            0.019474            0.254083        
Samsung Browser     10.0             1.518987            1.736370        
Mobile Safari       9.0              0.292113            0.465954        
                    10.0             0.603700            0.702179        
Samsung Browser     9.0              0.038948            0.128259        
Edge                17.0             0.116845            0.180212        
Safari              9.0              0.019474            0.064941        
Mobile Chrome       74.0             0.292113            0.336883        
Chrome              76.0             0.097371            0.140436        
Opera               66.0             0.019474            0.059259        
Mobile Chrome       70.0             0.077897            0.116083        
Samsung Browser     11.0             0.214216            0.250024        
Chrome              75.0             0.389484            0.424554        
                    71.0             0.038948            0.068188        
Mobile Safari (App) 10.0             0.058423            0.086047        
Mobile Chrome       72.0             0.038948            0.058447        
Edge                16.0             0.019474            0.038965        
Firefox             71.0             0.038948            0.049518        
Safari              11.0             0.350536            0.356366        
Mobile Chrome       68.0             0.038948            0.044647        
Chrome              74.0             0.038948            0.040588        
                    70.0             0.038948            0.038965        
Mobile Safari (App) 9.0              0.019474            0.017859        
Mobile Safari       7.0              0.019474            0.009741        
Mobile Chrome       50.0             0.019474            0.007306        
Chrome              67.0             0.058423            0.045459        
Mobile Safari       8.0              0.038948            0.025977        
Firefox             60.0             0.019474            0.005682        
Mobile Chrome       43.0             0.019474            0.004871        
Firefox             56.0             0.019474            0.004059        
Mobile Chrome       45.0             0.019474            0.004059        
Facebook            254.0            0.019474            0.004059        
Firefox             65.0             0.019474            0.004059        
Mobile Chrome       65.0             0.019474            0.004059        
Chrome              42.0             0.019474            0.003247        
Firefox             53.0             0.019474            0.002435        
                    64.0             0.019474            0.002435        
Chrome              30.0             0.019474            0.002435        
IE                  11.0             0.019474            0.002435        
Mobile Chrome       58.0             0.019474            0.000812        
                    57.0             0.019474            0.000812        
Firefox             73.0             0.03

In [12]:
flagged_traffic_type_orders = pd.pivot_table(flagged_orders, index =["traffic_type"], values="order_flag", aggfunc="count")
not_flagged_traffic_type_orders =pd.pivot_table(non_flagged_orders, index =["traffic_type"], values="order_flag", aggfunc="count")

display(pd.pivot_table(flagged_orders, index =["order_flag"], values="date", aggfunc=[max, min]))
display(pd.pivot_table(non_flagged_orders, index =["order_flag"], values="date", aggfunc=[max, min]))

traffic_type_check_orders = flagged_traffic_type_orders.join(not_flagged_traffic_type_orders, lsuffix="_fl", rsuffix="_not") 

display(traffic_type_check_orders)

traffic_type_check_orders.to_csv(r"C:\Users\level\Downloads\granifyproject-main\granifyproject-main\BA Candidate Assignment 1 Files [EXTERNAL]\order_discrepancy_traffic_type.csv")

max                  min
                           date                 date
order_flag                                          
1           2020-02-10 04:59:58  2020-02-07 05:02:44

max                  min
                           date                 date
order_flag                                          
0           2020-02-10 04:59:09  2020-02-07 05:00:03

order_flag_fl  order_flag_not
traffic_type                                
active         5070           123187.0      
ip_blacklist   58             1.0           
not_supported  7             NaN

In [13]:
print(flagged_orders.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5135 entries, 22 to 127199
Data columns (total 19 columns):
site_id            5135 non-null int64
order_id           5135 non-null int64
date               5135 non-null object
test_group         5010 non-null float64
campaign_id        2530 non-null float64
subtotal           5135 non-null float64
traffic_type       5135 non-null object
inserted           5135 non-null object
os                 5135 non-null object
os_version         4434 non-null object
browser            5135 non-null object
browser_version    4800 non-null float64
device_category    5135 non-null object
ssid               4997 non-null object
Segment_ck         481 non-null object
Revenue_ck         481 non-null float64
revenue_flag       5135 non-null int32
segment_num        5135 non-null int32
order_flag         5135 non-null int32
dtypes: float64(5), int32(3), int64(2), object(9)
memory usage: 902.2+ KB
None


In [14]:
flagged_browser_check = pd.pivot_table(flagged_orders, index =['os','os_version','browser', "browser_version"], values="order_flag", aggfunc=func1)
combined_browser_check = pd.pivot_table(non_flagged_orders, index =['os', 'os_version', 'browser', "browser_version"], values="order_flag", aggfunc=func2)

browser_check = flagged_browser_check.join(combined_browser_check, lsuffix="_flagged", rsuffix="_nf")
browser_check["diff"] = browser_check["order_flag_flagged"] - browser_check["order_flag_nf"]
display(browser_check.sort_values("diff"))

order_flag_flagged  \
os      os_version browser             browser_version                       
iOS     13         Mobile Safari       13.0             7.283350             
Mac     OS X       Safari              13.0             7.575463             
Windows 10         Edge                18.0             4.926972             
iOS     13         Mobile Safari (App) 13.0             13.106134            
Windows 10         Chrome              80.0             3.076923             
iOS     13         Mobile Chrome       54.0             0.603700             
Android 8          Mobile Chrome       79.0             1.635833             
iOS     13         Mobile Chrome       79.0             0.564752             
Mac     OS X       Safari              12.0             0.214216             
iOS     11         Mobile Safari       11.0             0.486855             
Android 7          Mobile Chrome       79.0             0.798442             
                                       80.0             0.370010             
iOS     9          Mobile Safari       9.0              0.292113             
Mac     OS X       Chrome              79.0             0.662123             
Windows 7          Chrome              80.0             0.545278             
Android 6          Mobile Chrome       80.0             0.077897             
Windows 8.1        Chrome              79.0             0.389484             
iOS     10         Mobile Safari       10.0             0.603700             
Android 8          Mobile Chrome       78.0             0.077897             
Windows 10         Edge                17.0             0.116845             
Android 5          Mobile Chrome       79.0             0.038948             
iOS     12         Mobile Chrome       54.0             0.155794             
Mac     OS X       Safari              9.0              0.019474             
Android 5          Samsung Browser     10.0             0.019474             
        8          Mobile Chrome       77.0             0.038948             
iOS     10         Mobile Safari (App) 10.0             0.058423             
Android 7          Samsung Browser     10.0             0.194742             
                   Mobile Chrome       78.0             0.058423             
Mac     OS X       Chrome              49.0             0.019474             
Windows 10         Opera               66.0             0.019474             
                   Edge                16.0             0.019474             
Android 6          Mobile Chrome       78.0             0.019474             
        5          Mobile Chrome       78.0             0.019474             
        7          Mobile Chrome       75.0             0.019474             
        6          Samsung Browser     10.0             0.097371             
        8          Mobile Chrome       70.0             0.077897             
Mac     OS X       Chrome              78.0             0.038948             
Android 8          Mobile Chrome       75.0             0.058423             
Mac     OS X       Safari              11.0             0.350536             
Windows 10         Chrome              75.0             0.272639             
Android 5          Mobile Chrome       80.0             0.038948             
        7          Mobile Chrome       76.0             0.019474             
        4.4        Mobile Chrome       79.0             0.019474             
        5          Mobile Chrome       77.0             0.019474             
iOS     9          Mobile Safari (App) 9.0              0.019474             
Windows XP         Chrome              49.0             0.077897             
        10         Chrome              76.0             0.019474             
Android 4.4        Chrome              79.0             0.058423             
        5          Chrome              80.0             0.038948             
        7          Mobile Chrome       74.0             0.019474             


In [15]:
print(flagged_orders["device_category"].value_counts() / len(flagged_orders.index))
print(non_flagged_orders["device_category"].value_counts() / len(non_flagged_orders.index))

mobile     0.437196
desktop    0.415385
tablet     0.134761
unknown    0.012658
Name: device_category, dtype: float64
mobile     0.495901
desktop    0.382578
tablet     0.121513
unknown    0.000008
Name: device_category, dtype: float64


In [16]:
flagged_browser_check = pd.pivot_table(flagged_orders, index =['browser', "browser_version",'os','os_version'], values="order_flag", aggfunc=func1)
combined_browser_check = pd.pivot_table(non_flagged_orders, index =['browser', "browser_version",'os','os_version'], values="order_flag", aggfunc=func2)

browser_check = flagged_browser_check.join(combined_browser_check, lsuffix="_flagged", rsuffix="_nf")
browser_check["diff"] = browser_check["order_flag_flagged"] - browser_check["order_flag_nf"]
display(browser_check.sort_values("diff"))

order_flag_flagged  \
browser             browser_version os      os_version                       
Mobile Safari       13.0            iOS     13          7.283350             
Safari              13.0            Mac     OS X        7.575463             
Edge                18.0            Windows 10          4.926972             
Mobile Safari (App) 13.0            iOS     13          13.106134            
Chrome              80.0            Windows 10          3.076923             
Mobile Chrome       54.0            iOS     13          0.603700             
                    79.0            Android 8           1.635833             
                                    iOS     13          0.564752             
Safari              12.0            Mac     OS X        0.214216             
Mobile Safari       11.0            iOS     11          0.486855             
Mobile Chrome       79.0            Android 7           0.798442             
                    80.0            Android 7           0.370010             
Mobile Safari       9.0             iOS     9           0.292113             
Chrome              79.0            Mac     OS X        0.662123             
                    80.0            Windows 7           0.545278             
Mobile Chrome       80.0            Android 6           0.077897             
Chrome              79.0            Windows 8.1         0.389484             
Mobile Safari       10.0            iOS     10          0.603700             
Mobile Chrome       78.0            Android 8           0.077897             
Edge                17.0            Windows 10          0.116845             
Mobile Chrome       79.0            Android 5           0.038948             
                    54.0            iOS     12          0.155794             
Safari              9.0             Mac     OS X        0.019474             
Samsung Browser     10.0            Android 5           0.019474             
Mobile Chrome       77.0            Android 8           0.038948             
Mobile Safari (App) 10.0            iOS     10          0.058423             
Samsung Browser     10.0            Android 7           0.194742             
Mobile Chrome       78.0            Android 7           0.058423             
Chrome              49.0            Mac     OS X        0.019474             
Opera               66.0            Windows 10          0.019474             
Edge                16.0            Windows 10          0.019474             
Mobile Chrome       78.0            Android 6           0.019474             
                                            5           0.019474             
                    75.0            Android 7           0.019474             
Samsung Browser     10.0            Android 6           0.097371             
Mobile Chrome       70.0            Android 8           0.077897             
Chrome              78.0            Mac     OS X        0.038948             
Mobile Chrome       75.0            Android 8           0.058423             
Safari              11.0            Mac     OS X        0.350536             
Chrome              75.0            Windows 10          0.272639             
Mobile Chrome       80.0            Android 5           0.038948             
                    76.0            Android 7           0.019474             
                    79.0            Android 4.4         0.019474             
                    77.0            Android 5           0.019474             
Mobile Safari (App) 9.0             iOS     9           0.019474             
Chrome              49.0            Windows XP          0.077897             
                    76.0            Windows 10          0.019474             
                    79.0            Android 4.4         0.058423             
                    80.0            Android 5           0.038948             
Mobile Chrome       74.0            Android 7           0.019474             


In [17]:
display(pd.pivot_table(flagged_orders, index =['browser', "browser_version",'os','os_version'], values="order_flag", aggfunc='count'))
#combined_browser_check = pd.pivot_table(non_flagged_orders, index =['browser', "browser_version",'os','os_version'], values="order_flag", aggfunc=func2)


order_flag
browser             browser_version os      os_version            
Chrome              30.0            Android 4.4         1         
                    42.0            Android 4.2         1         
                    49.0            Mac     OS X        1         
                                    Windows Vista       10        
                                            XP          4         
                    67.0            Mac     OS X        3         
                    70.0            Android 7           1         
                                            8           1         
                    71.0            Android 5           1         
                                    Windows 10          1         
                    74.0            Windows 7           1         
                    75.0            Windows 10          14        
                                            7           3         
                    76.0            Android 5           1         
                                    Windows 10          1         
                    77.0            Windows 10          3         
                                            8.1         2         
                    78.0            Android 5           3         
                                            7           1         
                                            8           1         
                                    Mac     OS X        2         
                                    Windows 10          10        
                    79.0            Android 4.4         3         
                                            5           10        
                                            6           8         
                                            7           24        
                                            8           16        
                                    Mac     OS X        34        
                                    Windows 10          349       
                                            7           93        
                                            8           3         
                                            8.1         20        
                    80.0            Android 5           2         
                                            6           3         
                                            7           15        
                                            8           14        
                                    Mac     OS X        13        
                                    Windows 10          158       
                                            7           28        
                                            8           5         
                                            8.1         21        
                    81.0            Windows 10          4         
Edge                14.0            Windows 10          3         
                    15.0            Windows 10          2         
                    16.0            Windows 10          1         
                    17.0            Windows 10          6         
                    18.0            Windows 10          253       
Facebook            253.0           Android 8           1         
                    254.0           Android 8           1         
Firefox             48.0            Windows 10          2         
                    52.0            Windows Vista       8         
                                            XP          6         
                    53.0            Windows 10          1         
                    56.0            Windows 10          1         
                    60.0            Windows 10          1         
                    61.0            Windows 7           1         
                    63.0            Mac     OS X        2         
                    64.0            Mac     OS X        1         
                    65.0            

Seems like order discrepancies are disproportionately coming from Firefox version 72 across operating systems, though primarily on Windows 7 and 10 as those are the most common OS. 

NameError: name 'flagged_firefox72' is not defined

Alright, looking for patterns among the blacklisted IPs.

In [46]:
flagged_blacklist = flagged_orders[(flagged_orders["traffic_type"] == "ip_blacklist")]

display(flagged_blacklist.head())

site_id   order_id                 date  test_group  campaign_id  \
4219  1402     367522335  2020-02-07 05:04:01 NaN         NaN            
6404  1402     367540011  2020-02-07 13:57:09 NaN         NaN            
7178  1402     367540095  2020-02-07 14:15:53 NaN         NaN            
8186  1402     367538543  2020-02-07 13:46:56 NaN         NaN            
8936  1402     367544127  2020-02-07 15:10:51 NaN         NaN            

      subtotal  traffic_type             inserted       os os_version  \
4219  35.99     ip_blacklist  2020-02-08 23:49:40  Windows  10          
6404  18.12     ip_blacklist  2020-02-08 23:49:40  Windows  10          
7178  16.80     ip_blacklist  2020-02-08 23:49:40  iOS      13          
8186  21.36     ip_blacklist  2020-02-08 23:49:40  Windows  10          
8936  263.40    ip_blacklist  2020-02-08 23:49:40  Windows  10          

                  browser  browser_version device_category ssid Segment_ck  \
4219  Chrome               79.0             unknown         NaN  NaN         
6404  Chrome               79.0             unknown         NaN  NaN         
7178  Mobile Safari (App)  13.0             unknown         NaN  NaN         
8186  Chrome               79.0             unknown         NaN  NaN         
8936  Chrome               80.0             unknown         NaN  NaN         

      Revenue_ck  revenue_flag  segment_num  order_flag  
4219 NaN          1             1            1           
6404 NaN          1             1            1           
7178 NaN          1             1            1           
8186 NaN          1             1            1           
8936 NaN          1             1            1

In [44]:
flagged_not_supported = flagged_orders[(flagged_orders["traffic_type"] == "not_supported")]

display(flagged_not_supported.head())

site_id   order_id                 date  test_group  campaign_id  \
36497  1402     367613129  2020-02-08 04:04:30 NaN         NaN            
46241  1402     367625645  2020-02-08 05:12:23 NaN         NaN            
46242  1402     367626087  2020-02-08 05:12:23 NaN         NaN            
46243  1402     367627787  2020-02-08 05:12:23 NaN         NaN            
57582  1402     367632189  2020-02-08 06:29:46 NaN         NaN            

       subtotal   traffic_type             inserted       os os_version  \
36497  27.60     not_supported  2020-02-09 23:51:53  Windows  Phone       
46241  38.99     not_supported  2020-02-09 23:51:53  Windows  Phone       
46242  38.99     not_supported  2020-02-09 23:51:53  Windows  Phone       
46243  34.99     not_supported  2020-02-09 23:51:53  Windows  Phone       
57582  107.22    not_supported  2020-02-09 23:51:53  iOS      3           

             browser  browser_version device_category ssid Segment_ck  \
36497  other         NaN               unknown         NaN  NaN         
46241  other         NaN               unknown         NaN  NaN         
46242  other         NaN               unknown         NaN  NaN         
46243  other         NaN               unknown         NaN  NaN         
57582  Mobile Safari  4.0              unknown         NaN  NaN         

       Revenue_ck  revenue_flag  segment_num  order_flag  
36497 NaN          1             1            1           
46241 NaN          1             1            1           
46242 NaN          1             1            1           
46243 NaN          1             1            1           
57582 NaN          1             1            1

After playing around with the data looking for patterns, it seems 

In [45]:
display(flagged_orders.sort_values("ssid").head())

site_id   order_id                 date  test_group  campaign_id  \
9053    1402     367541129  2020-02-07 14:25:07  0.0         6733.0        
89659   1402     367700435  2020-02-09 00:43:04  0.0         6230.0        
124244  1402     367799273  2020-02-10 04:42:25  0.0         6190.0        
74894   1402     367667673  2020-02-08 18:33:47  0.0        NaN            
71310   1402     367700055  2020-02-09 00:39:38  0.0        NaN            

        subtotal traffic_type             inserted       os os_version  \
9053    29.95     active       2020-02-08 23:49:40  Android  7           
89659   39.90     active       2020-02-10 23:50:57  iOS      13          
124244  59.99     active       2020-02-11 23:57:04  iOS      11          
74894   43.99     active       2020-02-09 23:51:53  Android  5           
71310   17.95     active       2020-02-10 23:50:57  Android  NaN         

                    browser  browser_version device_category  \
9053    other               NaN               tablet           
89659   Mobile Safari (App)  13.0             mobile           
124244  Mobile Safari (App)  11.0             mobile           
74894   other               NaN               tablet           
71310   Mobile Chrome        80.0             mobile           

                                                        ssid Segment_ck  \
9053    00000080-3018-4800-98e8-70b858281088:1402:1581085507  NaN         
89659   00000080-6030-4808-8800-008080402090:1402:1581208984  NaN         
124244  00000080-70b8-4828-a870-3898e8f0f8f8:1402:1581309745  NaN         
74894   00008040-2090-4860-a070-b8d848201008:1402:1581186827  NaN         
71310   00008040-b0d8-4870-9868-b0d8a8d0e870:1402:1581208778  NaN         

        Revenue_ck  revenue_flag  segment_num  order_flag  
9053   NaN          1             1            1           
89659  NaN          1             1            1           
124244 NaN          1             1            1           
74894  NaN          1             1            1           
71310  NaN          1             1            1

In [29]:
flagged_revenue = combined_data[(combined_data["revenue_flag"] == 1) & (combined_data["order_flag"] == 0)]
non_flagged_revenue = combined_data[combined_data["revenue_flag"] == 0]

flagged_both = combined_data[(combined_data["revenue_flag"] == 1) & (combined_data["order_flag"] == 1)]
flagged_revenue_no_order_disc = combined_data[(combined_data["revenue_flag"] == 0) & (combined_data["order_flag"] == 1)]

In [30]:
print(flagged_revenue.info())
print(flagged_revenue_no_order_disc.info())
print(flagged_both.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 477 entries, 249 to 127194
Data columns (total 19 columns):
site_id            477 non-null int64
order_id           477 non-null int64
date               477 non-null object
test_group         477 non-null float64
campaign_id        206 non-null float64
subtotal           477 non-null float64
traffic_type       477 non-null object
inserted           477 non-null object
os                 477 non-null object
os_version         414 non-null object
browser            477 non-null object
browser_version    449 non-null float64
device_category    477 non-null object
ssid               475 non-null object
Segment_ck         14 non-null object
Revenue_ck         14 non-null float64
revenue_flag       477 non-null int32
segment_num        477 non-null int32
order_flag         477 non-null int32
dtypes: float64(5), int32(3), int64(2), object(9)
memory usage: 68.9+ KB
None
<class 'pandas.core.frame.DataFrame'>
Int64Index: 481 entries, 596 to 126

In [31]:
func1 = lambda x: 100*x.count()/flagged_revenue.shape[0]
func2 = lambda x: 100*x.count()/non_flagged_revenue.shape[0]

flagged_os_check = pd.pivot_table(flagged_revenue, index =['os', "os_version"], values="revenue_flag", aggfunc=func1)
combined_os_check = pd.pivot_table(non_flagged_revenue, index =['os', "os_version"], values="revenue_flag", aggfunc=func2)

os_check = flagged_os_check.join(combined_os_check, lsuffix="_flagged", rsuffix="_nf")
os_check["diff"] = os_check["revenue_flag_flagged"] - os_check["revenue_flag_nf"]
display(os_check.sort_values("diff"))


revenue_flag_flagged  revenue_flag_nf      diff
os      os_version                                                 
iOS     13          19.287212             27.078057       -7.790845
Android 7           2.725367              3.329762        -0.604395
iOS     9           0.209644              0.496786        -0.287142
        10          0.628931              0.833658        -0.204727
        11          0.838574              1.036593        -0.198019
Android 6           1.048218              1.052828        -0.004610
Windows 8           0.209644              0.086856         0.122787
        8.1         1.257862              1.020358         0.237503
Android 4.4         0.838574              0.288980         0.549595
iOS     12          7.127883              6.556432         0.571450
Android 8           5.031447              4.318462         0.712984
        5           4.402516              2.778590         1.623926
Windows 7           4.612159              2.427106         2.185054
Mac     OS X        15.513627             13.306059        2.207568
Windows 10          23.060797             19.745600        3.315196

In [34]:
flagged_browser_check = pd.pivot_table(flagged_revenue, index =['browser', "browser_version"], values="revenue_flag", aggfunc=func1)
combined_browser_check = pd.pivot_table(non_flagged_revenue, index =['browser', "browser_version"], values="revenue_flag", aggfunc=func2)

browser_check = flagged_browser_check.join(combined_browser_check, lsuffix="_flagged", rsuffix="_nf")
browser_check["diff"] = browser_check["revenue_flag_flagged"] - browser_check["revenue_flag_nf"]
display(browser_check.sort_values("diff"))

browser_check.to_csv(r"C:\Users\level\Downloads\granifyproject-main\granifyproject-main\BA Candidate Assignment 1 Files [EXTERNAL]revenue_discrepancy_browsers.csv")

revenue_flag_flagged  revenue_flag_nf  \
browser             browser_version                                          
Mobile Safari (App) 13.0             10.482180             14.487142         
Mobile Safari       13.0             7.127883              10.604585         
Mobile Chrome       79.0             9.224319              10.967433         
Edge                18.0             6.289308              7.350315          
Mobile Chrome       80.0             5.241090              6.204948          
Mobile Safari (App) 12.0             1.257862              2.146243          
Mobile Chrome       54.0             0.419287              1.254140          
Samsung Browser     10.0             1.048218              1.733879          
Chrome              79.0             10.272537             10.854601         
Safari              12.0             0.209644              0.496786          
Mobile Safari       9.0              0.209644              0.465939          
Mobile Chrome       77.0             0.209644              0.387200          
Safari              11.0             0.209644              0.357166          
Mobile Safari       11.0             0.628931              0.763848          
Chrome              80.0             5.241090              5.340444          
Mobile Safari       10.0             0.628931              0.699721          
Mobile Safari (App) 11.0             0.209644              0.230534          
Chrome              77.0             0.209644              0.146113          
Samsung Browser     9.0              0.209644              0.127443          
Mobile Chrome       74.0             0.419287              0.335249          
Chrome              71.0             0.209644              0.067375          
Mobile Chrome       72.0             0.209644              0.057634          
Firefox             68.0             0.209644              0.035717          
Chrome              81.0             0.209644              0.009741          
Firefox             53.0             0.209644              0.001623          
Edge                17.0             0.419287              0.178583          
Chrome              78.0             0.628931              0.253263          
Samsung Browser     11.0             0.628931              0.247581          
Mobile Chrome       78.0             1.257862              0.717579          
Chrome              75.0             1.257862              0.420482          
Safari              13.0             11.111111             10.169492         
Mobile Safari       12.0             5.450734              4.021365          
Firefox             72.0             12.578616             2.388954          

                                          diff  
browser             browser_version             
Mobile Safari (App) 13.0            -4.004962   
Mobile Safari       13.0            -3.476702   
Mobile Chrome       79.0            -1.743114   
Edge                18.0            -1.061007   
Mobile Chrome       80.0            -0.963858   
Mobile Safari (App) 12.0            -0.888382   
Mobile Chrome       54.0            -0.834853   
Samsung Browser     10.0            -0.685661   
Chrome              79.0            -0.582064   
Safari              12.0            -0.287142   
Mobile Safari       9.0             -0.256296   
Mobile Chrome       77.0            -0.177557   
Safari              11.0            -0.147522   
Mobile Safari       11.0            -0.134917   
Chrome              80.0            -0.099354   
Mobile Safari       10.0            -0.070790   
Mobile Safari (App) 11.0            -0.020891   
Chrome              77.0             0.063530   
Samsung Browser     9.0              0.082200   
Mobile Chrome       74.0             0.084038   
Chrome              71.0             0.142269   
Mobile Chrome       72.0             0.152010   
Firefox             68.0             0.173927   
Chrome              81.0             0.199903   
Firefox             53.0        

In [27]:
display(pd.pivot_table(flagged_revenue, index =["traffic_type"], values="order_flag", aggfunc="count"))
display(pd.pivot_table(non_flagged_revenue, index =["traffic_type"], values="order_flag", aggfunc="count"))

order_flag
traffic_type            
active        477

order_flag
traffic_type            
active        123190    
ip_blacklist  2